pkg/apiserver/proxy_test.go

```diff
@@ -34,6 +36,115 @@ import (
 	"k8s.io/kubernetes/pkg/api/rest"
 )
 
+func calculateContentLength(reqBody string, transferEncodings []string) int64 {
+	if len(transferEncodings) == 0 || (len(transferEncodings) == 1 && transferEncodings[0] == "identity") {
+		return int64(len(reqBody))
+	}
+	// RFC2616 section 4.4: The Content-Length header field MUST NOT be sent if these two lengths
+	// are different (i.e., if a Transfer-Encoding header field is present).
+	return int64(-1)
+}
+
+func containGzip(transferEncodings []string) bool {
+	for _, encoding := range transferEncodings {
+		if encoding == "gzip" {
+			return true
+		}
+	}
+	return false
+}
+
+func TestProxyRequestContentLengthAndTransferEncoding(t *testing.T) {
+	serverResponse := "got response"
+	table := []struct {
+		transferEncodings []string
+		reqBody           string
+		reqNamespace      string
+	}{
+		{[]string{}, "", "default"},
+		{[]string{"identity"}, "question", "default"},
+		{[]string{"chunked"}, "question", "default"},
+		// RFC2616 section-3.6: Whenever a transfer-coding is applied to a message-body, the set of
+		// transfer-codings MUST include "chunked", unless the message is terminated by closing the
+		// connection.
+		{[]string{"chunked", "gzip"}, "qqqqqqqqqquestion", "default"},
+	}
+
+	for _, item := range table {
+		proxyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			expectedContentLength := calculateContentLength(item.reqBody, item.transferEncodings)
+			if e, a := expectedContentLength, req.ContentLength; e != a {
+				t.Errorf("expected %v, got %v", e, a)
+			}
+			// We expect Content-Length header field be set
+			if expectedContentLength != -1 {
+				contentLengthInHeader, err := strconv.Atoi(req.Header.Get("Content-Length"))
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if e, a := expectedContentLength, int64(contentLengthInHeader); e != a {
+					t.Errorf("expected %v, got %v", e, a)
+				}
+			}
+
+			// Help wanted: why is the transfer-encoding header field always empty?
```

Comment:
I prints the content in the buffer in the [conn](https://golang.org/src/net/http/server.go#L131) struct and [persistConn](https://golang.org/src/net/http/transport.go#L818) struct to understand what's written/read by the go standard library to/from the wire. The golang library only supports `chuncked` as a transfer-encoding, more specifically,

- When a request is sent either by a default golang http client or by a golang server, the only possible value in the transfer-encoding field is `chunked`,
https://golang.org/src/net/http/transfer.go#L178

- When a golang server reads the next request, the standard library parses the transfer-encoding field, passes the result to `Request.TransferEncoding`, and deletes the transfer-encoding field in the header, 
https://golang.org/src/net/http/transfer.go#L433,
The standard library actually considers it’s an error if the transfer-encoding field contains value other than `identity` or `chunked`,
https://golang.org/src/net/http/transfer.go#L447